### PR TITLE
[BUU] Delete Button missing before saving variant

### LINKED ISSUE #12400: [BUU] Delete button missing, before saving a variant
## Description

When we create a line for a new variant (by clicking the New Variant button), on the Backoffice UI Uplift (`admin_style_v3`) design, there is no possibility to delete this new line. The delete option appears only after saving the variant.

This means, that if you hit the button too many times, then you won't be able to save changes. You'll have to navigate away/refresh page, which can mean the info added to some variants will be lost.

## Expected Behavior

The delete variant button is visible before and after the variant is saved.

## Actual Behaviour

The delete variant button is visible only after the variant is saved.

## Steps to Reproduce

First enable [feature toggle](https://github.com/openfoodfoundation/openfoodnetwork/wiki/Feature-toggles) `admin_style_v3`

1. Visit the products page
2. Create a product
3. Hit the add variant button, to add a variant -> notice there is no delete button.

## Animated Gif/Screenshot

Legacy design:

(Note that you can also view the legacy screen in the new design, by visiting the shortcut `/admin/products_old`)
image

Current design:

image

## Workaround

Page refresh or just navigating away.

## Severity

bug-s4: it's annoying, but you can use it

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by cyrillefr**
#### What? Why?

- Closes #12400 

 I used the remove option of `stimulus-components/rails-nested-form` nodejs lib
I added the Remove option in the menu

#### What should we test?

- Toggle to the v3 admin
- Visit `http://localhost:3000/admin/products`
- Chose some product and click the blue cross under the last variant line
- At the end of line, the three little dots menu should appear
- It should have only one option: 'Remove'
- Clicking 'Remove' should remove the line without confirmation

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [X] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Review by rioug — APPROVED** [review]
Nice one :+1: Thanks @cyrillefr

**Review by mkllnk — APPROVED** [review]

File: app/views/admin/products_v3/_variant_row.html.haml [hunk 1]
```diff
@@ -48,11 +48,15 @@
...
+    - if variant.persisted?
       = link_to t('admin.products_page.actions.edit'), edit_admin_product_variant_path(variant.product, variant)
       - if variant.product.variants.size > 1
         %a{ "data-controller": "modal-link", "data-action": "click->modal-link#setModalDataSetOnConfirm click->modal-link#open",
             "data-modal-link-target-value": "variant-delete-modal", "class": "delete",
             "data-modal-link-modal-dataset-value": {'data-current-id': variant.id}.to_json }
           = t('admin.products_page.actions.delete')
+    - else
+      %a{ 'data-action': "nested-form#remove" }
+        = f.hidden_field :_destroy
```
Comment (dacook) [inline app/views/admin/products_v3/_variant_row.html.haml:61]: I hadn't seen a field inside an `<a>` before. I guess it works if specs pass!

Comment (cyrillefr) [reply app/views/admin/products_v3/_variant_row.html.haml]: Aah, it works but it is not valid html. I have updated the haml and put the hidden field out of the anchor.

**Comment by cyrillefr** [comment]
Hello @sigmundpetersen , 
I would need a rerun of the checks :) 
A locally unknown error appeared

**Comment by mkllnk** [comment]
There's a conflicting file. This needs a rebase.

**Comment by dacook** [comment]
Hi @cyrillefr.

I tested this, and have noticed a couple of things to fix. Can you please try and fix these?

1. I think the colour of the "remove" text should be red, like the "delete" action. Although it's not deleting a record, it's still destructive because it deletes anything that you had typed in to the new line.
Screenshot 2024-05-08 at 10 30 17 am

2. After removing a new variant, it still seems to be hidden in the background. It still says "product modified", and when you save, it still tries to save it.
Screenshot 2024-05-08 at 10 27 46 am

I've also rebased it to fix the conflict.

File: spec/system/admin/products_v3/products_spec.rb [hunk 2]
```diff
@@ -530,6 +530,20 @@
...
+        new_variant_row = find_field("Name", placeholder: "Apples", with: "").ancestor("tr")
+        within new_variant_row do
+          fill_in "Name", with: "Large box"
+          fill_in "SKU", with: "APL-02"
+          expect(page).to have_field("Name", placeholder: "Apples", with: "Large box")
+          page.find(".vertical-ellipsis-menu").click
+          page.find('a', text: 'Remove').click
+        end
+
+        expect(page).not_to have_field("Name", placeholder: "Apples", with: "Large box")
```
Comment (dacook) [inline spec/system/admin/products_v3/products_spec.rb:544]: Also expect not to have "product modified", and able to proceed with saving a change on an existing record.

**Review by dacook — CHANGES_REQUESTED** [review]
PS can

**Review by dacook — COMMENTED** [review]
Thanks for adding more specs!

I had some ideas to simplify it so did an experiment.
1. I thought that stimulus-rails-nested-form should delete the DOM elements for us, so I investigated and found that we were missing an attribute required for this (see my first commit).
2. I also thought that instead of rebuilding the recordElements array, we could simply ignore elements that aren't connected to the DOM (see second commit).
3. Then I found another case to test for! (third commit)

What do you think of this approach, would you agree?

**Comment by dacook** [comment]
Rebased on master to fix conflicts.

**Comment by cyrillefr** [comment]
Hello @dacook ,

Of course I agree :)

Brilliant use of the `isConnected` method that I have never used ...
I have missed the new record/existing record , the doc could be clearer but I also never practiced that kind of nested form.

In my opinion, I think this is fine now.

**Review by rioug — APPROVED** [review]

**Review by dacook — APPROVED** [review]

**Comment by dacook** [comment]
Thanks @cyrillefr , I didn't think of that case at first either, that's fine, sometimes things like this only come up with manual testing.
And t hanks to your work on the specs, I was able to add the new case quite easily 👍

**Comment by filipefurtad0** [comment]
Hey @cyrillefr ,

Thanks for fixing this.

It looks great: after clicking the Add Variant button, we can immediately see the three dots which contain the new Remove option.
Clicking it, removes the yet to be saved variant: the new button works as expected.

image

It's worth noting that if two variants are added (and not saved), and one of them is removed, the page is checked for changes and the banner appears, stating there are unsaved changes:

image

I think this is an edge case and I'm not sure this is even an issue as, strictly speaking, there are indeed unsaved changes - it seems correct to me. Using the Remove button for the remaining variant also removes the banner, so I think we're good here.

Thanks!

**Pull request merged**